lightningd/watchtower.c

```diff
@@ -0,0 +1,127 @@
+#include <bitcoin/tx.h>
+#include <common/htlc_tx.h>
+#include <common/key_derive.h>
+#include <common/keyset.h>
+#include <hsmd/gen_hsm_wire.h>
+#include <lightningd/channel.h>
+#include <wire/wire_sync.h>
+
+static const u8 ONE = 0x1;
+
+const struct bitcoin_tx *
+penalty_tx_create(const tal_t *ctx, struct lightningd *ld,
+		  const struct channel *channel,
+		  const struct secret *revocation_preimage,
+		  const struct bitcoin_txid *commitment_txid,
+		  s16 to_them_outnum, struct amount_sat to_them_sats)
+{
+	u8 *wscript;
+	struct bitcoin_tx *tx;
+	struct keyset keyset;
+	size_t weight;
+	u8 *msg;
+	struct amount_sat fee, min_out, amt;
+	struct bitcoin_signature sig;
+	u32 locktime = 0;
+	bool option_static_remotekey = channel->option_static_remotekey;
+	struct pubkey final_key;
+	u8 **witness;
+	u32 remote_to_self_delay = channel->channel_info.their_config.to_self_delay;
+	const struct amount_sat dust_limit = channel->our_config.dust_limit;
+	u32 feerate_per_kw = try_get_feerate(ld->topology, FEERATE_PENALTY);
+	BUILD_ASSERT(sizeof(struct secret) == sizeof(*revocation_preimage));
+	const struct secret remote_per_commitment_secret = *revocation_preimage;
+	struct pubkey remote_per_commitment_point;
+	struct basepoints basepoints[NUM_SIDES];
+	basepoints[LOCAL] = channel->local_basepoints;
+	basepoints[REMOTE] = channel->channel_info.theirbase;
+
+	if (to_them_outnum == -1 ||
+	    amount_sat_less_eq(to_them_sats, dust_limit)) {
+		log_unusual(channel->log,
+			    "Cannot create penalty transaction because there "
+			    "is no non-dust to_them output in the commitment.");
+		return NULL;
+	}
+
+	if (!pubkey_from_secret(&remote_per_commitment_secret, &remote_per_commitment_point))
+		fatal("Failed derive from per_commitment_secret %s",
+		      type_to_string(tmpctx, struct secret,
+				     &remote_per_commitment_secret));
+
+	if (!bip32_pubkey(ld->wallet->bip32_base, &final_key,
+			  channel->final_key_idx)) {
+		fatal("Could not derive onchain key %" PRIu64,
+		      channel->final_key_idx);
+	}
+
+	if (!derive_keyset(&remote_per_commitment_point,
+			   &basepoints[REMOTE],
+			   &basepoints[LOCAL],
+			   option_static_remotekey,
+			   &keyset))
+		abort(); /* TODO(cdecker) Handle a bit more gracefully */
+	wscript = bitcoin_wscript_to_local(tmpctx, remote_to_self_delay,
+					   &keyset.self_revocation_key,
+					   &keyset.self_delayed_payment_key);
+
+	tx = bitcoin_tx(ctx, chainparams, 1, 1, locktime);
+	bitcoin_tx_add_input(tx, commitment_txid, to_them_outnum, 0xFFFFFFFF,
+			     to_them_sats, NULL);
+
+	bitcoin_tx_add_output(tx, scriptpubkey_p2wpkh(tx, &final_key),
+			      to_them_sats);
+
+	/* Worst-case sig is 73 bytes */
+	weight = bitcoin_tx_weight(tx) + 1 + 3 + 73 + 0 + tal_count(wscript);
+	weight = elements_add_overhead(weight, 1, 1);
+	fee = amount_tx_fee(feerate_per_kw, weight);
+
+	/* Result is trivial?  Spend with small feerate, but don't wait
+	 * around for it as it might not confirm. */
```

Comment:
this portion of this comment doesn't make sense out of context, since it's referring to how onchaind tracks output resolution

>but don't wait around for it as it might not confirm